### DISCUSSION TITLE
Containerization

### DISCUSSION THREAD

**Discussion by NonSwag** [PaperMC/Paper#10597] [Ideas]
### Is your feature request related to a problem?

This is not really a problem, rather an inconvenience to deal with...
I am using docker to containerize my servers and always face the same issues when updating plugins, operators, whitelist, bans...
Due to the nature of docker volumes, it is not possible to save single files to a volume which are not inside a folder.
When updating plugins, I have to update the volume and the image.

### Describe the solution you'd like.

So my proposal is to add a feature to change the file structure in a way that makes it possible to include those files in a volume.
For example, for the plugins' problem it would be great to have a folder where the plugins are loaded from, but a different folder for the data.
In case of the caches, libs and version jars a parent folder would be perfect so the jars are not "folderless" in the volumes (if joined).
This could be a startup flag or an option in the paper global config.
To preserve consistency it would make sense to add startup flags for defining the path of the bans-, ops-, whitelist -files just like for the server.properties, bukkit- and commands.yml files.

### Describe alternatives you've considered.

Currently, I am dealing with this by updating everything in both the volume and the image, which is a pain in the a$$.
For not having to deal with the lack of persistent files for whitelist, operators and bans I wrote a hacky plugin that allows me to save those in a volume without storing the entire image.

**Comment by lynxplay** [comment]
I am pretty sure you can configure the `settings.world-container` in the bukkit.yml to move your "data" from the rest of your server data ?

What specifically are you looking to move into its own volume.

**Comment by NonSwag** [comment]
I already use the world container option, this is not my problem.
I want to move the other data away from the "persistent" files: plugin data, whitelist, operators, (ip-)bans
In general, there should be nothing within the volume that is not user specific

**Comment by electronicboy** [comment]
It's the general, we want to do something about this, but,
1) plugin authors have generally not uniformly been good at programming stuff in a good manner here, i.e. hardcoding their plugin data folder vs using the getDataFolder() method
2) any time we've talked about cleaning up the folder structure, we've generally had many people against it, which while there seems to be more for than against this stuff, it creates the headaches over motivation for this sort of work

**Comment by NonSwag** [comment]
I see your point, but going the step and making it optional would probably encourage developer to adapt to the new possibilities in the future

**Comment by XxGoldenbluexX** [comment]
Does this changed since the hard fork?

**Reply by Astralchroma** [reply]
Not substantially.

While Paper is now in a position to dictate how things work; the practicality of doing something still has to be considered, and this has not changed.
